server/audit_math/suite.py

```diff
@@ -1,718 +0,0 @@
-            if np.any(np.array(pvalues) == 0):
-                fisher_pvalues[i] = 0
-            else:
-                obs = -2 * np.sum(np.log(pvalues))
-                fisher_pvalues[i] = 1 - sp.stats.chi2.cdf(obs, df=2 * len(pvalues))
-
-        pvalue = np.max(fisher_pvalues)
-        alloc_lambda = test_lambdas[np.argmax(fisher_pvalues)]
-
-        # If p-value is over the risk limit, then there's no need to refine the
-        # maximization. We have a lower bound on the maximum.
-        if pvalue > alpha or modulus is None:
-            maximized_pvalue = pvalue
-            break
-
-        # Use modulus of continuity for the Fisher combination function to check
-        # how close this is to the true max
-        fisher_fun_obs = sp.stats.chi2.ppf(1 - pvalue, df=4)
-        fisher_fun_alpha = sp.stats.chi2.ppf(1 - alpha, df=4)
-        dist = np.abs(fisher_fun_obs - fisher_fun_alpha)
-        mod = modulus(stepsize)
-
-        if mod <= dist:
-            maximized_pvalue = pvalue
-            break
-
-        # We haven't found a good enough max yet, keep looking
-        # TODO memoize the p-values we've already looked at
-        # to make it faster.
-        lambda_lower = alloc_lambda - 2 * stepsize
-        lambda_upper = alloc_lambda + 2 * stepsize
-        stepsize /= 10
-
-    return min(maximized_pvalue, 1.0)
-
-
-def try_n(
-    n: int,
-    alpha: float,
-    contest: Contest,
-    winner: str,
-    loser: str,
-    bp_stratum: BallotPollingStratum,
-    cvr_stratum: BallotComparisonStratum,
-    n_ratio: float,
-) -> float:
-    """
-    A function that evaluates whether a certain sample size can meet the risk limit.
-
-    Inputs:
-        n - the sample size to test
-        alpha - the risk limit as a fraction
-        contest - the contest being audited
-        winner, loser - the winner-loser pair to evaluate the risk measurement over
-        bp_stratum: a stratum object containing the ballot polling stratum
-        cvr_stratum: a stratum object containing the ballot comparison stratum
-        n_ratio: the ratio of the size of the CVR stratum to both strata.
-    Outputs:
-        A Fisher's combined maximized p-value for the given sample size, contest,
-        winner-loser pair, and strata.
-    """
-    n1_original = cvr_stratum.sample_size
-    n2_original = bp_stratum.sample_size
-
-    o1_rate, o2_rate, u1_rate, u2_rate = 0.0, 0.0, 0.0, 0.0
-    # Assume o1, o2, u1, u2 rates will be the same as what we observed in sample
-    if n1_original != 0:
-        o1_rate = cvr_stratum.misstatements[(winner, loser)]["o1"] / n1_original
-        o2_rate = cvr_stratum.misstatements[(winner, loser)]["o2"] / n1_original
-        u1_rate = cvr_stratum.misstatements[(winner, loser)]["u1"] / n1_original
-        u2_rate = cvr_stratum.misstatements[(winner, loser)]["u2"] / n1_original
-
-    n1 = math.ceil(n_ratio * n)
-    n2 = int(n - n1)
-
-    if (n1 < n1_original) or (n2 < n2_original):
-        return 1.0
```

Comment:
i didn't see this case in the original code. curious why you needed to add it

Reply:
It's here: https://github.com/pbstark/CORLA18/blob/master/code/suite_tools.py#L616

Basically if our new sample sizes (which should be cumulative and include previous rounds) are lower than the ones we started with, something has gone terrible wrong.  Typically it happens during the linear search, so this just makes sure we get back to where we're supposed to be.